### PR TITLE
Treat intersection types as potential `never`

### LINKED ISSUE #46032: Crash trying to assing undefined[] to never[]
# Bug Report

### 🔎 Search Terms
`never[] crash` `generalized source shouldn't be assignable` `never union`

### 🕗 Version & Regression Information

In 4.0 starts to crash, still happens in the nightly build.

To reproduce the crash, create a file and paste the code, the simply compile it with 

```
tsc file.ts
```

- [x] This is a crash
- This changed between versions 3.9 and 4.0

### 💻 Code

```ts
interface A {
  foo: "a";
}

interface B {
  foo: "b";
}

type Union = A & B;

function func(): { value: Union[] } {
  return {
    value: [],
  };
}
```

### 🙁 Actual behaviour

Crashes with error

```
Error: Debug Failure. False expression: generalized source shouldn't be assignable
    at reportRelationError (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:47173:30)
    at reportErrorResults (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:47368:25)
    at isRelatedTo (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:47329:17)
    at typeArgumentsRelatedTo (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:47570:39)
    at relateVariances (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:47961:34)
    at structuredTypeRelatedTo (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:47911:46)
    at recursiveTypeRelatedTo (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:47653:53)
    at isRelatedTo (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:47308:38)
    at checkTypeRelatedTo (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:46997:26)
    at elaborateElementwise (/Users/eliancordoba/ts-crash/node_modules/typescript/lib/tsc.js:46407:38)
```

### 🙂 Expected behaviour

Either show an error message or compile successfully

### PR DESCRIPTION
**Pull request by lowr**
Fixes #46032

This PR fixes crashes and improves error messages when a function is declared to return a type alias that names an intersection type of object-like types that reduces to `never`.

Besides #46032, there have been certain circumstances under which using such type aliases as functions' return type emits error messages different from using `never` directly ([playground](https://www.typescriptlang.org/play?strict=false&noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&ts=4.6.0-dev.20220116#code/C4TwDgpgBAglC8UDeUwC4oCIaagXwG4AoUSKAIQWVQ03N0JPGgGEq4AyC4ogMwFcAdgGNgASwD2gqLwCMACgCUGQRABuEAE7IiUKJojB+mwcTxE+Q0ZOm8ATEoxsku-YeOmi5yyPFSoAOYKylCqGtpI3gK+NoEOIc54QA)). This PR improves error messages so that they emit the same errors (aside from elaboration of `never` intersections). This is actually not really related to #46032, so let me know if I should open a separate issue/PR.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -17815,6 +17815,8 @@ namespace ts {
             const t = target.flags;
             if (t & TypeFlags.AnyOrUnknown || s & TypeFlags.Never || source === wildcardType) return true;
             if (t & TypeFlags.Never) return false;
+            // Intersection types that would be reduced to never may be passed as `target`, thus returning early.
+            if (t & TypeFlags.Intersection) return false;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:17819]: This isn't quite the right fix, I will put something up in a PR.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -32138,7 +32140,8 @@ namespace ts {
             }
 
             const functionFlags = getFunctionFlags(func);
-            const type = returnType && unwrapReturnType(returnType, functionFlags);
+            const unwrappedType = returnType && unwrapReturnType(returnType, functionFlags);
+            const type = unwrappedType && getReducedType(unwrappedType);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:32144]: I think this is best covered in a separate PR since it is really an unrelated fix.

**Review by ahejlsberg — CHANGES_REQUESTED** [review]

**Comment by ahejlsberg** [comment]
@lowr Thanks for your work on this issue. My suggested fix is in #47816. Feel free to copy from that into this PR, or we can just merge mine.

File: src/compiler/checker.ts (hunk 1 above)
Comment (lowr) [reply src/compiler/checker.ts:17819]: Thanks for the review. Besides my fix not allowing empty array literals assigned to `never[]`, is it not "quite the right fix" because it fails to catch cases where `target` is a union of empty intersections that also reduce to `never`? A bit of explanation or even pointers would be appreciated if you don't mind!

Comment (ahejlsberg) [reply src/compiler/checker.ts:17819]: Yes, one issue is that both unions and intersections can potentially reduce to `never`, so both need to be excluded when `--strictNullChecks` is disabled. But another issue is that `any` should be assignable to anything, which is handled by a test later in the body of `isSimpleTypeRelatedTo`. So the code can't just unconditionally return `false` here.

Comment (lowr) [reply src/compiler/checker.ts:17819]: Thanks! I'll think about those logic more carefully next time.
